cleanlab/datalab/internal/issue_manager/underperf_group.py

```diff
@@ -126,24 +127,84 @@ def set_knn_graph(self, features: npt.NDArray, find_issues_kwargs: Dict) -> csr_
+        # ALTERNATIVE FORMULATION: Connect the graph using a dummy node before HDBSCAN.
+        # Adapted from - https://github.com/scikit-learn-contrib/hdbscan/issues/82#issuecomment-1242604639
+        # knn_graph = self.connect_knn_graph(knn_graph)
+        cluster_labels = self.run_hdbscan_on_components(knn_graph)
+
+        return cluster_labels[:200]  # Slicing applicable if connect_knn_graph is called
+
+    def run_hdbscan_on_components(
+        self, knn_graph: csr_matrix, min_cluster_size: int = 5, **hdbscan_params: Dict[str, Any]
+    ) -> npt.NDArray[np.int_]:
+        n_components, component_labels = sp.csgraph.connected_components(
+            csgraph=knn_graph, directed=False, return_labels=True
+        )
+        all_clusters = np.empty(knn_graph.shape[0], dtype=object)
+        offset = 0
+        for component_idx in range(n_components):
+            component_mask = component_labels == component_idx
+            component_subgraph = knn_graph[component_mask, :][:, component_mask]
+            knn_subgraph = self._symmetrize_knn_graph(component_subgraph)
+            clusterer = HDBSCAN(metric="precomputed")
+            component_clusters = clusterer.fit_predict(knn_subgraph)
+            outliers_mask = np.in1d(component_clusters, self.OUTLIER_LABELS)
+            component_clusters[~outliers_mask] += offset
+            all_clusters[component_mask] = component_clusters
+            offset = np.max(component_clusters) + 1
+        return all_clusters
+
+    def connect_knn_graph(
+        self, knn_graph: csr_matrix, new_node_dist: Optional[float] = None
+    ) -> csr_matrix:
+        """
+        This function takes in a sparse graph (csr_matrix) that has more than
+        one component (multiple unconnected subgraphs) and appends another
+        node to the graph that is weakly connected to all other nodes.
+
+        Args:
+            knn_graph (scipy.sparse.csr_matrix):
+                Sparse graph with multiple components.
+                See scipy.sparse.csgraph.connected_components
+            new_node_dist (float):
+                Value to use for the connection strengh to all other nodes.
+                Value will be appended as elements in a new row and column at
+                the ends of the distance matrix.
+
+        Returns:
+            connected_knn_graph (scipy.sparse.csr_matrix):
+                Sparse graph with only one component.
+        Source: https://github.com/scikit-learn-contrib/hdbscan/issues/82#issuecomment-1242604639
+        """
+        N = knn_graph.shape[0]
+        if new_node_dist is None:
+            new_node_dist = (knn_graph.max() - knn_graph.min()) * 1000
+        connected_knn_graph = knn_graph.copy()
+        connected_knn_graph = sp.vstack((connected_knn_graph, np.full((1, N), new_node_dist)))
+        connected_knn_graph = sp.hstack((connected_knn_graph, np.full((N + 1, 1), new_node_dist)))
+        return connected_knn_graph.tocsr()
```

Comment:
Consider moving this to a utility module as a separate function in the near future.